keras_nlp/models/falcon/falcon_transformer_decoder.py

```diff
@@ -0,0 +1,249 @@
+            epsilon=self.layer_norm_epsilon,
+            dtype=self.dtype_policy,
+            name="input_layernorm",
+        )
+        self._input_layernorm.build(decoder_sequence_shape)
+
+        # Attention layers.
+        self.key_dim = self.hidden_dim // self.num_attention_heads
+        self._attention_layer = FalconAttention(
+            num_heads=self.num_attention_heads,
+            attention_dropout=self.attention_dropout,
+            dtype=self.dtype_policy,
+            name="attention",
+        )
+        self._attention_layer.build(
+            decoder_sequence_shape,
+        )
+
+        self._attention_dropout = keras.layers.Dropout(
+            rate=self.attention_dropout,
+            dtype=self.dtype_policy,
+            name="attention_dropout",
+        )
+
+        self._post_attention_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+            dtype=self.dtype_policy,
+            name="post_attention_layernorm",
+        )
+        self._post_attention_layernorm.build(decoder_sequence_shape)
+
+        # Feedforward layers.
+        # TODO: use_bias should be an argument to the transformer to support
+        # other sizes of models, e.g. 7B, that don't use bias.
+        self._dense_h_to_4h = keras.layers.Dense(
+            self.intermediate_dim,
+            activation=keras.activations.gelu,
+            use_bias=True,
+            dtype=self.dtype_policy,
+            name="dense_h_to_4h",
+        )
+        self._dense_h_to_4h.build(decoder_sequence_shape)
+
+        self._dense_4h_to_h = keras.layers.Dense(
+            self.hidden_dim,
+            use_bias=True,
+            dtype=self.dtype_policy,
+            name="dense_4h_to_h",
+        )
+        self._dense_4h_to_h.build(
+            (
+                decoder_sequence_shape[0],
+                decoder_sequence_shape[1],
+                self.intermediate_dim,
+            )
+        )
+
+        self._feedforward_dropout = keras.layers.Dropout(
+            rate=self.feedforward_dropout,
+            dtype=self.dtype_policy,
+            name="feedforward_dropout",
+        )
+
+        self.built = True
+
+    def call(
+        self,
+        inputs,
+        decoder_padding_mask=None,
+        decoder_attention_mask=None,
+        attention_cache=None,
+        attention_cache_update_index=None,
+        training=None,
+    ):
+        attention_mask = self._compute_attention_mask(
+            decoder_sequence=inputs,
+            decoder_padding_mask=decoder_padding_mask,
+            decoder_attention_mask=decoder_attention_mask,
+            attention_cache=attention_cache,
+            attention_cache_update_index=attention_cache_update_index,
+        )
+
+        residual = inputs
+
+        x = self._input_layernorm(inputs)
+
+        alibi = self._build_alibi_tensor(
+            self.num_attention_heads, decoder_padding_mask
+        )
```

Comment:
@mattdangerw Right now, I'm calculating alibi in every layer although it doesn't change from layer to layer! It would be more efficient to just calculate it once in the backbone but backbone init doesn't know the shapes yet! If you have any suggestions to be able to calculate the alibi once, let me know.

Reply:
Discussed this offline. This is fine for now because it doesn't seem a really compute intensive part. Ideally, it would be better to analyze to see what's the impact of this repetition and avoid repetition if it has a significant impact on the performance of the model.